sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java

```diff
@@ -0,0 +1,328 @@
+    DataflowAssert.that(triggered)
+        .inFinalPane(window)
+        .containsInAnyOrder(1, 2, 3, 4, 5);
+    DataflowAssert.that(triggered)
+        .inOnTimePane(window)
+        .containsInAnyOrder(1, 2, 3);
+    DataflowAssert.that(count)
+        .inWindow(window)
+        .satisfies(new SerializableFunction<Iterable<Long>, Void>() {
+          @Override
+          public Void apply(Iterable<Long> input) {
+            for (Long count : input) {
+              assertThat(count, allOf(greaterThanOrEqualTo(3L), lessThanOrEqualTo(5L)));
+            }
+            return null;
+          }
+        });
+    DataflowAssert.that(sum)
+        .inWindow(window)
+        .satisfies(new SerializableFunction<Iterable<Integer>, Void>() {
+          @Override
+          public Void apply(Iterable<Integer> input) {
+            for (Integer sum : input) {
+              assertThat(sum, allOf(greaterThanOrEqualTo(6), lessThanOrEqualTo(15)));
+            }
+            return null;
+          }
+        });
+
+    p.run();
+  }
+
+  @Test
+  public void testProcessingTimeTrigger() {
+    TestStream<Long> source = TestStream.create(
+        VarLongCoder.of())
+        .addElements(TimestampedValue.of(1L, new Instant(1000L)),
+            TimestampedValue.of(2L, new Instant(2000L)))
+        .advanceProcessingTime(Duration.standardMinutes(12))
+        .addElements(TimestampedValue.of(3L, new Instant(3000L)))
+        .advanceProcessingTime(Duration.standardMinutes(6))
+        .advanceWatermarkToInfinity();
+
+    Pipeline p = getPipeline();
+    PCollection<Long> sum = p.apply(source)
+        .apply(Window.<Long>triggering(AfterWatermark.pastEndOfWindow()
+            .withEarlyFirings(AfterProcessingTime.pastFirstElementInPane()
+                .plusDelayOf(Duration.standardMinutes(5)))).accumulatingFiredPanes()
+            .withAllowedLateness(Duration.ZERO))
+        .apply(Sum.longsGlobally());
+
+    DataflowAssert.that(sum).inEarlyGlobalWindowPanes().containsInAnyOrder(3L, 6L);
+
+    p.run();
+  }
+
+  @Test
+  public void testDiscardingMode() {
+    TestStream<String> stream =
+        TestStream.create(StringUtf8Coder.of())
+            .advanceWatermarkTo(new Instant(0))
+            .addElements(
+                TimestampedValue.of("firstPane", new Instant(100)),
+                TimestampedValue.of("alsoFirstPane", new Instant(200)))
+            .addElements(TimestampedValue.of("onTimePane", new Instant(500)))
+            .advanceWatermarkTo(new Instant(1001L))
+            .addElements(
+                TimestampedValue.of("finalLatePane", new Instant(750)),
+                TimestampedValue.of("alsoFinalLatePane", new Instant(250)))
+            .advanceWatermarkToInfinity();
+
+    Pipeline p = getPipeline();
+    FixedWindows windowFn = FixedWindows.of(Duration.millis(1000L));
+    Duration allowedLateness = Duration.millis(5000L);
+    PCollection<String> values =
```

Comment:
A comment on why the given pipeline actually tests discarding mode. Specifically, here we're testing discarding mode because we trigger multiple times, and if we were in accumulating mode, the elements in the first trigger firing would be duplicated.